src/compas/robots/model/base.py

```diff
@@ -0,0 +1,116 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from compas.files.urdf import URDFElement
+from compas.files.urdf import URDFGenericElement
+from compas.geometry import Frame
+
+__all__ = [
+    'FrameProxy',
+    'ProxyObject',
+]
+
+
+def _parse_floats(values):
+    return [float(i) for i in values.split()]
+
+
+def _generic_from_data_or_data(data):
+    try:
+        data = URDFGenericElement.from_data(data)
+    finally:
+        return data
+
+
+def _attr_to_data(attr):
+    return {k: v.data if hasattr(v, 'data') else v for k, v in attr.items()}
+
+
+def _attr_from_data(data):
+    return {k: _generic_from_data_or_data(d) for k, d in data.items()}
+
+
+class ProxyObject(object):
+    """Base proxy class to create object proxies.
+
+    An object proxy wraps the proxied object and forwards all calls to it
+    unless a method/attribute/property is found first on the proxy itself.
+
+    The magic part is ``__getattr__`` ."""
+
+    def __init__(self, obj):
+        self._proxied_object = obj
+
+    def __getattr__(self, attr):
+        return getattr(self._proxied_object, attr)
+
+    @classmethod
+    def create_proxy(cls, obj):
+        """Creates a proxy wrapping around an object, only if it's not already proxied."""
+        if obj and not isinstance(obj, cls):
+            return cls(obj)
+        return obj
+
+
+class FrameProxy(ProxyObject):
+    """Proxy class that adds URDF functionality to an instance of :class:`Frame`.
+
+    This class is internal and not intended to be referenced externally.
+    """
+
+    def get_urdf_element(self):
+        attributes = {
+            'xyz': "{} {} {}".format(self.point.x, self.point.y, self.point.z),
+            'rpy': "{} {} {}".format(*self.euler_angles())
+        }
+        return URDFElement('origin', attributes)
+
+    @classmethod
+    def from_urdf(cls, attributes, elements, text):
+        """Create origin instance from an URDF element.
+
+        Parameters
+        ----------
+        attributes : dict[str, Any]
+            Attributes of the URDF element.
+        elements: list[object]
+            Children elements of the URDF element.
+        text: str | None
```

Comment:
if `text=None` then this would be `text : str, optional`

Reply:
;) yep, of course 🤦